### DISCUSSION TITLE
Is there a simple way to make the model slower but smarter?

### DISCUSSION THREAD

**Discussion by araleza** [oobabooga/text-generation-webui#4667] [Q&A] [answered]
Hello, LLM n00b here...

I've started experimenting with the LLM models here, and they're amazing.  I know a lot more about generative image AIs, and I know they have ways to make them slower, but produce higher quality images, such as by increasing the 'iterations' for noise removal.

Is there a similar slider for LLM models in oobabooga?  I'm getting instantaneous responses, so wouldn't mind slightly slower replies and an IQ boost, if that's a thing that can be done with a few typical changes.

**Comment by HiroseKoichi** [comment, marked as answer]
There's three main factors that determines the quality of output you get: the number of parameters, quantization method, and sampling method.

When comparing two models of the same architecture, such as LLama-2-7B and LLama-2-13B, the one with the higher number of parameters will always be better than the one with lower parameters. This is not always the case when comparing different architectures, and a very popular model, Mistral-7B, generally outperforms LLama-2-13B despite being a lower parameter model. 13B models are kind of in a rough spot right now since Mistral outperforms it with a native 8k context window as opposed to LLama-2-13B's 4k native context window, so there's pretty much three tiers of parameters that people use at the moment: 7B, 34B, and 70B. The 7Bs are various fine-tunes of Mistral-7B, and the 34Bs are fine-tunes of YI, but I don't know much about what 70Bs are good since I don't use them. I recommend using the highest tier of model you can comfortably run, then finding a fine-tuned version of it that you like.

There are four main quantization methods: GGUF, GPTQ, AWQ, and EXLLAMA. GPTQ, AWQ, and EXLLAMA are quantization methods that only run on the GPU, while GGUF can balance the load between the CPU and GPU. GPTQ is the standard for running on GPU only, while AWQ is supposed to be an improved version of GPTQ, but I don't know much about EXLLAMA since it's still new and I personally use GGUF. Quantization is a way of improving performance at the cost of quality, but that doesn't mean you should always run an unquantized model if you want the best quality. There are different levels of quantization, and the least quantized version of all the methods usually only has a negligible dip in quality compared to FP-16.

I exclusively run GGUF models, so you'll have to look elsewhere for more details on the other quantization methods. Q_8 is essentially identical to FP-16, but with greatly reduced VRAM consumption. Q_6 has a slight dip in quality but further reduces the VRAM requirements. Q_4_K_M is a good mix between quality and VRAM requirements, and this is generally what you would use when running larger models. If you have to drop below Q_3, then it might be a better idea to use a lower parameter model, but technically, a higher parameter model even when using the highest quantization level will outperform a lower parameter model. I also don't recommend running a 7B model with anything other than Q_8 since lower parameter models seem more sensitive to quantization. [TheBloke](https://huggingface.co/TheBloke) is an absolute chad who creates quantized versions of just about any model you could want, which conveniently doubles as a model release news feed.

The sampling method also plays a big part in how the model responds. There is no definitively correct way to sample at the moment, so you'll have to play around with it to find what you prefer. There are two presets that I have found to be very good: tail-free sampling = 0.9 with top_a = 0.2 being optional, or min_p = 0.05. The repetition penalty is also useful since they tend to repeat once they approach their maximum context length, with some being more affected than others. I generally only use presence penalty = 1, or no repetition penalty at all, but if you want to use the standard repetition penalty, then I recommend a value of 1.15.

One more thing to consider is the context window that the model is trained on. This doesn't necessarily affect the quality of the output; it just allows the model to retain more of the conversation. There is no downside to increasing the context size, except for increased VRAM usage, so you should generally use the highest context size that your hardware allows. LLAMA-2 has a trained context window of 4K across the board, Mistral-7B has a trained context window of 8K, and YI has a trained context window of 200K. Increasing the context size can greatly increase the VRAM requirements, so you'll likely only be able to use up to 8K unless you have really good hardware or you're using a cloud service. You can still use a higher context size even if the model isn't trained on that length, but you'll have to use things like rope scaling to maintain coherency outside of the model's pre-trained context length, but keep in mind that it does cause a slight degradation in quality.

One more thing I forgot to mention is fine-tunes. A fine-tune is when you take a base model, such as LLAMA-2-13B or Mistral-7B, and train it on an instruction dataset. This can greatly improve the model's performance on various tasks depending on what's in the fine-tune dataset, and a fine-tune is pretty much always better than the base model. When using fine-tunes, it's very important that you use the chat template that it was trained on, or else you'll get sub-par output. You can find the chat template for each model on TheBloke's download page for each one.

**Reply by redthista** [reply]
This is super helpful! thanks for the info.

**Comment by araleza** [comment]
That's a great writeup, @HiroseKoichi, thanks!  I think a lot of people will find what you said there in web searches and learn a lot from what you've said.
